### PR TITLE
New EMS calling point, functions, and variables

### LINKED ISSUE #8184: Allow weather actuators to work with EMS
Issue overview
--------------
In E+ v9.3 and earlier the weather data actuators such as "Direct Solar" do not work for EMS as there is no appropriate calling point available.

AT: I’ve tried replacing BeginTimestepBeforePredictor with BeginZoneTimestepBeforeInitHeatBalance in the program calling manager, but the results are still the same. Very high spikes of direct normal solar are reported with corresponding high gains through East windows.

MW: *** Hmmm, looking more closely at the code, these two callingn points occur after ManageWeather which is where the weather actuators fire. The only calling point I see before ManageWeather is BeginNewEnvironment which is too early (and it's before the day/hour/timestep loops).
I'm not sure how these weather actuators were intended to work. Oh, here's a clue from the description in the EMS application guide "These provide the ability to alter weather data and were originally requested for use with ExternalInterface for using measured data." Sure enough, there's a call to ExternalInterfaceExchangeVariables just before ManageWeather. So, as it stands now. you can't do this with EMS, only ExternalInterface.
I could add a new EMS calling point easy enough for v9.4. BeginTimeStepBeforeManageWeather. If that would help, please post a new issues asking for this.

### Details
Heldesk ticket: [EnergyPlus #15517]: How to modify weather data using EMS

### Checklist
The attached zip file includes an idf and epw file for testing. The epw file includes some very high direct normal values which we are trying to filter out. 

After the fix I would hope that the reported incident solar on the 1m2 window **Block1:Zone1_Wall_2_0_0_0_0_0_Win** will be no higher than 1000W. Without the fix the solar goes over 4,500W. 

Also the reported site **Direct Solar** values should be no higher than 900 W/m2 (if EMS is allowed to change the Environment reported values)

- [x] *Here's a revised defect file that uses the new calling point and new functions and one with no EMS to show that the values noted above are too high without the EMS correction.* 
[8184-WeatherActuator.zip](https://github.com/NREL/EnergyPlus/files/5092575/8184-WeatherActuator.zip)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #8184 
 - Add new EMS calling point BeginZoneTimestepBeforeSetCurrentWeather
 - Add new EMS functions to access Today* and Tomorrow* weather data arrays.
 - Add new EMS built-in variables for TimeStepNum and TimeStepsPerHour (adding these because the existing Minute variable is unreliable)

ToDo
----------
- [x] Finish the complete set of functions
- [x] Finish documentation
- [x] Unit tests
- [ ] Post new issue about Minute built-in variable

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] ~If any diffs are expected, author must demonstrate they are justified using plots and descriptions~
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions - see #8184 for test files.
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] ~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~
 - [ ] ~If structural output changes, add to output rules file and add OutputChange label~
 - [ ] ~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataGlobals.cc
Comment (mjwitte) [inline src/EnergyPlus/DataGlobals.cc:166]: New calling point `BeginZoneTimestepBeforeSetCurrentWeather`

File: src/EnergyPlus/DataRuntimeLanguage.cc
Comment (mjwitte) [inline src/EnergyPlus/DataRuntimeLanguage.cc:181]: New functions to access `Today*` and `Tomorrow*` weather data arrays. (And, yeah, an enum class would be great here, but didn't want to get into those weeds here.)

File: src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc
Comment (mjwitte) [inline src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc:310]: `ManagerWeather` needs the full `state` now because of the new EMS call.

File: src/EnergyPlus/PluginManager.cc
Comment (mjwitte) [inline src/EnergyPlus/PluginManager.cc:1277]: Attempted to add the new calling point on thePython side. Untested.

File: src/EnergyPlus/RuntimeLanguageProcessor.cc
Comment (mjwitte) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:155]: New ERL built-in variables.

File: src/EnergyPlus/RuntimeLanguageProcessor.cc
Comment (mjwitte) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:1425]: Add new `@Today*` and `@Tomorrow*` ERL functions. And some reformatting of this giant `if` block.

File: src/EnergyPlus/RuntimeLanguageProcessor.cc
Comment (mjwitte) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:2461]: New functions all call the same common `TodayTomorrowWeather` function, passing in the appropriate data array to access.

File: src/EnergyPlus/RuntimeLanguageProcessor.cc
Comment (mjwitte) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:2556]: The new common function for the `@Today*` and `@Tomorrow*` ERL functions.

File: src/EnergyPlus/RuntimeLanguageProcessor.cc
Comment (mjwitte) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:2571]: Overload for Rain and Snow which are bools. The others are all reals.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:195]: Several arrays were unused, so removing them: `TodayRainAmount, TodaySnowAmount, TodayOutWetBulbTemp, TodayOutHumRat, TomorrowRainAmount, TomorrowSnowAmount.` If they become useful someday they can be added back.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:528]: Here's the actual new calling point. Calling this during sizing or before KickOffSimulation results in missing ERL variables, because of input timing issues. So, for now, this calling point won't fire during sizing (noted in the docs). As long as we rely on KickOffSimulation to fire every component for input processing, I don't see a way around this. Unless we move that up before sizing, but that could have bad side effects.

File: src/EnergyPlus/WeatherManager.hh
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.hh:158]: Just adding units here to the endline comments.

File: src/EnergyPlus/api/runtime.cc
Comment (mjwitte) [inline src/EnergyPlus/api/runtime.cc:118]: Monkey see, monkey do. Untested.

**Review by mjwitte — COMMENTED** [review]
Code walkthrough.

**Comment by mjwitte** [comment]
@Myoldmopar or @mitchute This is ready for review for I/O freeze.

**Comment by Myoldmopar** [comment]
CI is very happy, as expected.  And thanks for the walkthrough.  I'll need to add this to the Python side of things so that the callback is available in Python EMS and also in the Python API "call as a library" world.

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/PluginManager.cc:1277]: I'll test.  Thanks for putting this in place.

File: src/EnergyPlus/RuntimeLanguageProcessor.cc
Comment (Myoldmopar) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:2500]: Fixed typo tomorrow -> today

File: src/EnergyPlus/RuntimeLanguageProcessor.cc
Comment (Myoldmopar) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:2612]: I added my own worker functions for mining the specific weather data.  It's not 100% ideal, but it's like 90% ideal.  It's still really nice to pack this logic into reusable pieces.  I couldn't use the ones previously added because their job was specifically to update Erl operands, and I needed raw values.  No time to make it more elegant right now.

File: src/EnergyPlus/api/datatransfer.cc
Comment (Myoldmopar) [inline src/EnergyPlus/api/datatransfer.cc:663]: Added today/tomorrow weather functions to the C API here.

File: src/EnergyPlus/api/datatransfer.py
Comment (Myoldmopar) [inline src/EnergyPlus/api/datatransfer.py:136]: Point our API instance to the underlying C functions.

File: src/EnergyPlus/api/datatransfer.py
Comment (Myoldmopar) [inline src/EnergyPlus/api/datatransfer.py:1032]: Add Python wrappings around each API function to allow a client to call into this from either Python EMS or through raw API calls during a callback in a running simulation.

File: src/EnergyPlus/api/plugin.py
Comment (Myoldmopar) [inline src/EnergyPlus/api/plugin.py:65]: Add the new calling point to the Python EMS Plugin base class, allowing users to override the function and EnergyPlus will then call the overridden version in their derived class instance.

File: src/EnergyPlus/api/runtime.h
Comment (Myoldmopar) [inline src/EnergyPlus/api/runtime.h:259]: Add the function to register the callback via API here.

File: src/EnergyPlus/api/runtime.py
Comment (Myoldmopar) [inline src/EnergyPlus/api/runtime.py:311]: Expose the new API calling point to Python clients.

File: testfiles/PythonPluginCustomSchedule.py
Comment (Myoldmopar) [inline testfiles/PythonPluginCustomSchedule.py:13]: This demonstrates the new calling point being called in a Python Plugin example file, and also demonstrates calling into one of the new weather worker functions.

File: tst/EnergyPlus/api/TestDataTransfer.py
Comment (Myoldmopar) [inline tst/EnergyPlus/api/TestDataTransfer.py]: This demonstrates calling one of the new weather worker functions through the C API in a callback scenario.

**Review by Myoldmopar — APPROVED** [review]
I think this will be good to go now assuming CI is happy.

**Comment by Myoldmopar** [comment]
Looks great so far.  Spurious GPU diffs showing up on Linux, but otherwise clean.  Assuming debug and Windows both are clean this can merge.

**Comment by mjwitte** [comment]
@Myoldmopar Thanks for all the changes to flesh this out on the Python side. Do you need to add anything there for the two new built-in variables?

**Comment by Myoldmopar** [comment]
Oh.  I didn't realize those were also new.  Yep, incoming...

**Comment by Myoldmopar** [comment]
Done.  Now if it is clean we'll call this done.

**Comment by Myoldmopar** [comment]
I already pulled in develop locally and built/ran with the other API change that just went in.  It succeeded completely, no need to push and wait for yet another round.  The results from ad67002 will suffice unless something big changes.

**Comment by Myoldmopar** [comment]
Windows is bogged down it seems, so I'm building it locally on Windows with develop pulled in to check it out.  As soon as it finishes I'll merge this.

**Comment by Myoldmopar** [comment]
I see Windows starting here, but I'm going to merge this to get things moving.  I built it locally with no problem.  The unit test failed on what appears to be a totally unrelated thing, and maybe a fluke.  If it persists, I'll investigate.

**Pull request merged**
